### PR TITLE
[DocsDB] Change modify_placement_info of yb-admin to specify number of copies of each Cloud.Region.Zone

### LINKED ISSUE #10467: [DocDB] Change modify_placement_info of yb-admin to specify number of copies of each Cloud.Region.Zone
### Description

Currently, for e.g. say that out of 3 copies of a tablet, we want to have 2 copies in `C1.R1.Z1` and 1 copy in `C2.R2.Z2` then we would specify it as below:

```
./bin/yb-admin \
    -master_addresses $MASTER_RPC_ADDRS \
    modify_placement_info  \
    C1.R1.Z1,C1.R1.Z1,C2.R2.Z2 3
```

However, this is somewhat complex and hard to follow. We don't need to necessarily repeat `C1.R1.Z1` 2 times or any other placement block the number of times we want it to be replicated. Instead, the above example could become:

```
./bin/yb-admin \
    -master_addresses $MASTER_RPC_ADDRS \
    modify_placement_info  \
    C1.R1.Z1 2 C2.R2.Z2 1 3
```

This change is fairly simple and only involves changing the client interface. Underneath, even today we count the copies and store and use them, so server-side changes should be trivial.

### PR DESCRIPTION
**Pull request by NitronR**
Resolves #10467

### CODE REVIEW

File: src/yb/tools/yb-admin_client.cc [hunk 1]
```diff
@@ -1741,7 +1741,27 @@ Status ClusterAdminClient::ModifyPlacementInfo(
...
+    if(placement_info_min_replica_split.size() == 0 ||
+        placement_info_min_replica_split.size() > 2) {
+      return STATUS(InvalidCommand, "Each placement info must have at most 2 values separated by a colon. "
+          "Format: cloud.region.zone:[min_replica_count]. Invalid placement info: " + placement_info_split[iter]);
+    }
+
+    std::string placement_target = placement_info_min_replica_split[0],
+        placement_min_replica_count = "1";
+
+    if(placement_to_min_replicas.find(placement_target) != placement_to_min_replicas.end()) {
```
Comment (NitronR) [inline src/yb/tools/yb-admin_client.cc:1756]: In case a cloud.region.zone is mentioned twice, returing invalid command. Alternatives can be:
1. Final count can be total of all mentioned entries.
2. Use the first or last mentioned entry.

File: src/yb/tools/yb-admin-test.cc [hunk 2]
```diff
@@ -538,6 +538,21 @@ TEST_F(AdminCliTest, TestGetClusterLoadBalancerState) {
   ASSERT_NE(output.find("ENABLED"), std::string::npos);
 }
 
+TEST_F(AdminCliTest, TestModifyPlacementPolicy) {
+  BuildAndStart();
+
+  // Modify the cluster placement policy to consist of 2 zones.
+  ASSERT_OK(CallAdmin("modify_placement_info", "c.r.z0,c.r.z1:2", 2, ""));
```
Comment (sanketkedia) [inline src/yb/tools/yb-admin-test.cc:545]: Why 2 here? Does this test even work? I would assume we need a value >= 3 here. We have a validation check somewhere that the RF needs to be >= sum(min_num_replicas of all placement blocks)

**Comment by sanketkedia** [comment]
Hey @NitronR, Nice set of changes! Good work! Left some preliminary comments. I would also suggest adding screenshot of the outputs you get when you run the yb-admin commands in the terminal.

Also, for docs changes, the content is present in [yb-admin.md](https://github.com/yugabyte/yugabyte-db/blob/master/docs/content/latest/admin/yb-admin.md).  Feel free to reach out with questions.

**Review by sanketkedia — CHANGES_REQUESTED** [review]

File: src/yb/tools/yb-admin-test.cc (hunk 2 above)
Comment (sanketkedia) [reply src/yb/tools/yb-admin-test.cc:545]: Just checked the code, we don't have any such validation in the `modify_placement_info` code path. Can you add a check in the below function that the `total replica count <= num_replicas specified`?

Comment (NitronR) [reply src/yb/tools/yb-admin-test.cc:545]: Sure

File: docs/content/v2.0/admin/yb-admin.md [hunk 3]
```diff
@@ -534,18 +534,23 @@ yb-admin -master_addresses <master-addresses> modify_placement_info <placement_i
...
+- *replication_factor*: The number of replicas for each tablet. This value should be more than the total of replica counts specified in *placement_info*.
 - *placement_id*: The identifier of the primary cluster, which can be any unique string. If not set, a randomly-generated ID will be used.
 
 **Example**
 
 ```sh
 $ ./bin/yb-admin --master_addresses $MASTER_RPC_ADDRS \
     modify_placement_info  \
-    aws.us-west.us-west-2a,aws.us-west.us-west-2b,aws.us-west.us-west-2c 3
+    aws.us-west.us-west-2a:2,aws.us-west.us-west-2b:3,aws.us-west.us-west-2c,aws.us-west.us-west-2b:2 9
```
Comment (sanketkedia) [inline docs/content/v2.0/admin/yb-admin.md:546]: Let's not repeat a Placement block in the example since that would introduce more confusion. That is just for backward compatibility and we shouldn't encourage it to the user.

File: docs/content/v2.0/admin/yb-admin.md [hunk 4]
```diff
@@ -534,18 +534,23 @@ yb-admin -master_addresses <master-addresses> modify_placement_info <placement_i
 ```
 
 - *master-addresses*: Comma-separated list of YB-Master hosts and ports. Default value is `localhost:7100`.
-- *placement_info*: Comma-delimited list of placements for *cloud*.*region*.*zone*. Default value is `cloud1.datacenter1.rack1`.
-- *replication_factor*: The number of replicas for each tablet.
+- *placement_info*: Comma-delimited list of placements for *cloud*.*region*.*zone*. Default value is `cloud1.datacenter1.rack1`. Optionally, a minimum replica count can be specified after a colon, default being 1. If a placement is specified multiple times, the total count from all mentions are taken.
```
Comment (sanketkedia) [inline docs/content/v2.0/admin/yb-admin.md:537]: Rephrase the last line to:  It is not recommended to repeat placement multiple times but instead specify the total count after the colon. However, in the event that the user specifies a placement multiple times, the total count from all mentions is taken.

File: src/yb/tools/yb-admin_client.cc [hunk 5]
```diff
@@ -1738,10 +1738,34 @@ Status ClusterAdminClient::ModifyPlacementInfo(
...
+    }
+
+    std::string placement_target = placement_info_min_replica_split[0],
+        placement_min_replica_count = "1";
+
+    if(placement_info_min_replica_split.size() == 2) {
+      placement_min_replica_count = placement_info_min_replica_split[1];
+    }
+
+    int min_replica_count = VERIFY_RESULT(CheckedStoi(placement_min_replica_count));
```
Comment (sanketkedia) [inline src/yb/tools/yb-admin_client.cc:1762]: Nit: can combine converting the string to an integer in the `if` block above.

File: src/yb/tools/yb-admin_client.cc [hunk 6]
```diff
@@ -1738,10 +1738,34 @@ Status ClusterAdminClient::ModifyPlacementInfo(
...
+        placement_block, ":", strings::AllowEmpty());
+
+    if(placement_info_min_replica_split.size() == 0 ||
+        placement_info_min_replica_split.size() > 2) {
+      return STATUS(InvalidCommand, "Each placement info must have at most 2 values separated by a colon. "
+          "Format: cloud.region.zone:[min_replica_count]. Invalid placement info: " + placement_block);
+    }
+
+    std::string placement_target = placement_info_min_replica_split[0],
+        placement_min_replica_count = "1";
```
Comment (sanketkedia) [inline src/yb/tools/yb-admin_client.cc:1756]: Let's keep this variable of integer type and perform the type conversion in the `if` block below.

File: src/yb/tools/yb-admin_client.cc [hunk 7]
```diff
@@ -1738,10 +1738,34 @@ Status ClusterAdminClient::ModifyPlacementInfo(
...
+
+    if(placement_info_min_replica_split.size() == 2) {
+      placement_min_replica_count = placement_info_min_replica_split[1];
```
Comment (sanketkedia) [inline src/yb/tools/yb-admin_client.cc:1759]: Use CheckedStoi here itself and convert it to an integer.

File: src/yb/tools/yb-admin_client.cc [hunk 8]
```diff
@@ -1738,10 +1738,34 @@ Status ClusterAdminClient::ModifyPlacementInfo(
...
+    }
+
+    int min_replica_count = VERIFY_RESULT(CheckedStoi(placement_min_replica_count));
+    total_min_replica_count += min_replica_count;
+    placement_to_min_replicas[placement_target] += min_replica_count;
+  }
+
+  if (total_min_replica_count > replication_factor) {
+      return STATUS(InvalidCommand, "replication_factor should be more than the total of replica counts specified in placement_info.");
```
Comment (sanketkedia) [inline src/yb/tools/yb-admin_client.cc:1768]: num_replicas should be greater equal .... ?

**Review by sanketkedia — CHANGES_REQUESTED** [review]
@NitronR Left some more cosmetic comments. Functionally, the logic looks correct.

File: docs/content/v2.0/admin/yb-admin.md (hunk 3 above)
Comment (NitronR) [reply docs/content/v2.0/admin/yb-admin.md:546]: Makes sense, will update

**Comment by NitronR** [comment]
@sanketkedia Thanks for the suggestions. I have pushed the updates.

File: docs/content/v2.0/admin/yb-admin.md [hunk 9]
```diff
@@ -534,18 +534,23 @@ yb-admin -master_addresses <master-addresses> modify_placement_info <placement_i
 ```
 
 - *master-addresses*: Comma-separated list of YB-Master hosts and ports. Default value is `localhost:7100`.
-- *placement_info*: Comma-delimited list of placements for *cloud*.*region*.*zone*. Default value is `cloud1.datacenter1.rack1`.
-- *replication_factor*: The number of replicas for each tablet.
+- *placement_info*: Comma-delimited list of placements for *cloud*.*region*.*zone*. Default value is `cloud1.datacenter1.rack1`. It is not recommended to repeat placement multiple times but instead specify the total count after the colon. However, in the event that the user specifies a placement multiple times, the total count from all mentions is taken.
```
Comment (sanketkedia) [inline docs/content/v2.0/admin/yb-admin.md:537]: Why did you remove this line: `Optionally, a minimum replica count can be specified after a colon, the default being 1.` I had commented to rephrase only the last line. So let's add the above line before line:`It is not recommended to repeat placement multiple times...`

Also, remove the line: `Default value is cloud1.datacenter1.rack1` as that is no longer applicable. We expect at least 2 arguments i.e. the placement info and the RF.

File: docs/content/v2.0/admin/yb-admin.md [hunk 10]
```diff
@@ -534,18 +534,23 @@ yb-admin -master_addresses <master-addresses> modify_placement_info <placement_i
...
+- *replication_factor*: The number of replicas for each tablet. This value should be more than the total of replica counts specified in *placement_info*.
 - *placement_id*: The identifier of the primary cluster, which can be any unique string. If not set, a randomly-generated ID will be used.
 
 **Example**
 
 ```sh
 $ ./bin/yb-admin --master_addresses $MASTER_RPC_ADDRS \
     modify_placement_info  \
-    aws.us-west.us-west-2a,aws.us-west.us-west-2b,aws.us-west.us-west-2c 3
+    aws.us-west.us-west-2a:2,aws.us-west.us-west-2b:3,aws.us-west.us-west-2c 8
```
Comment (sanketkedia) [inline docs/content/v2.0/admin/yb-admin.md:546]: We generally add an RF value that is odd. Change `aws.us-west.us-west-2b:3` to `aws.us-west.us-west-2b:2` and the RF to 5.

File: src/yb/tools/yb-admin_client.cc [hunk 11]
```diff
@@ -1738,10 +1738,33 @@ Status ClusterAdminClient::ModifyPlacementInfo(
...
   live_replicas->set_num_replicas(replication_factor);
 
+  int total_min_replica_count = 0;
+
   // Iterate over the placement blocks of the placementInfo structure.
   std::unordered_map<std::string, int> placement_to_min_replicas;
-  for (int iter = 0; iter < placement_info_split.size(); iter++) {
-    placement_to_min_replicas[placement_info_split[iter]]++;
+  for (const auto& placement_block : placement_info_split) {
+    std::vector<std::string> placement_info_min_replica_split = strings::Split(
```
Comment (sanketkedia) [inline src/yb/tools/yb-admin_client.cc:1746]: Nit: Generally for complicated types we like to use `auto` to maintain readability. So, replace `std::vector<std::string>` with auto

File: docs/content/v2.0/admin/yb-admin.md [hunk 12]
```diff
@@ -534,18 +534,23 @@ yb-admin -master_addresses <master-addresses> modify_placement_info <placement_i
 ```
```
Comment (sanketkedia) [inline docs/content/v2.0/admin/yb-admin.md:534]: @NitronR this is the wrong file to make docs changes. I had commented to make changes in https://github.com/yugabyte/yugabyte-db/blob/master/docs/content/latest/admin/yb-admin.md (i.e. the latest version) and not on v2.0

File: docs/content/v2.0/admin/yb-admin.md [hunk 13]
```diff
@@ -534,18 +534,23 @@ yb-admin -master_addresses <master-addresses> modify_placement_info <placement_i
...
 ```sh
 $ ./bin/yb-admin --master_addresses $MASTER_RPC_ADDRS \
     modify_placement_info  \
-    aws.us-west.us-west-2a,aws.us-west.us-west-2b,aws.us-west.us-west-2c 3
+    aws.us-west.us-west-2a:2,aws.us-west.us-west-2b:3,aws.us-west.us-west-2c 8
 ```
 
+This will place a minimum of:
+1. 2 replicas in aws.us-west.us-west-2a
+2. 3 replicas in aws.us-west.us-west-2b
```
Comment (sanketkedia) [inline docs/content/v2.0/admin/yb-admin.md:551]: Update the count to 2

**Review by sanketkedia — CHANGES_REQUESTED** [review]
@NitronR Thanks for such a quick turnaround and your persistence. For docs changes, I think you are editing the wrong file. We need to change the latest version and not v2.0. Elaborated in the comment. Also, left final set of comments

Thanks!

File: docs/content/v2.0/admin/yb-admin.md (hunk 12 above)
Comment (NitronR) [reply docs/content/v2.0/admin/yb-admin.md:534]: Oh right, missed that. Thanks for pointing that out.

File: src/yb/tools/yb-admin_client.cc (hunk 11 above)
Comment (NitronR) [reply src/yb/tools/yb-admin_client.cc:1746]: Tried doing that but it did not work. If it is auto, the type becomes Splitter<..> which does not have size method or indexing. Specifying vector does type conversion Splitter has a converter for Container types.
```
src/yb/tools/yb-admin_client.cc:1749:41: error: ‘class strings::internal::Splitter<strings::delimiter::Literal, strings::AllowEmpty>’ has no member named ‘size’
     if(placement_info_min_replica_split.size() == 0 ||
```
Also in repo, I see similar approach of specifying the type instead of auto so this will keep it consistent with other code as well. Let me know your thoughts.

**Comment by NitronR** [comment]
@sanketkedia pushed the updates and replied to one of the comments. Please check when possible, thanks.

File: docs/content/latest/admin/yb-admin.md [hunk 14]
```diff
@@ -1008,8 +1008,8 @@ yb-admin \
 ```
 
 * _master-addresses_: Comma-separated list of YB-Master hosts and ports. Default value is `localhost:7100`.
-* *placement_info*: Comma-delimited list of placements for *cloud*.*region*.*zone*. Default value is `cloud1.datacenter1.rack1`.
-* *replication_factor*: The number of replicas for each tablet.
+* *placement_info*: Comma-delimited list of placements for *cloud*.*region*.*zone*. Optionally, a minimum replica count can be specified after a colon, the default being 1. It is not recommended to repeat placement multiple times but instead specify the total count after the colon. However, in the event that the user specifies a placement multiple times, the total count from all mentions is taken.
+* *replication_factor*: The number of replicas for each tablet. This value should be more than the total of replica counts specified in *placement_info*.
```
Comment (sanketkedia) [inline docs/content/latest/admin/yb-admin.md:1012]: Nit: Shouldn't the line be: `This value should be greater equal to the sum total of replica counts...`

File: docs/content/latest/admin/yb-admin.md [hunk 15]
```diff
@@ -1008,8 +1008,8 @@ yb-admin \
 ```
 
 * _master-addresses_: Comma-separated list of YB-Master hosts and ports. Default value is `localhost:7100`.
-* *placement_info*: Comma-delimited list of placements for *cloud*.*region*.*zone*. Default value is `cloud1.datacenter1.rack1`.
-* *replication_factor*: The number of replicas for each tablet.
+* *placement_info*: Comma-delimited list of placements for *cloud*.*region*.*zone*. Optionally, a minimum replica count can be specified after a colon, the default being 1. It is not recommended to repeat placement multiple times but instead specify the total count after the colon. However, in the event that the user specifies a placement multiple times, the total count from all mentions is taken.
```
Comment (sanketkedia) [inline docs/content/latest/admin/yb-admin.md:1011]: The first line still looks a little vague to me. Maybe rephrase to: `Optionally, after each placement block, we can also specify a minimum replica count separated by a colon. This count indicates how many minimum replicas of each tablet we want in that placement block. Its default value is 1. It is not recommended to repeat the same placement multiple times ...` and the remaining text?

**Review by sanketkedia — CHANGES_REQUESTED** [review]
@NitronR Your comment makes sense. So, ignore my previous comment then.

Just a couple more minor comments to improve the sentence structure in the docs file, otherwise, we are all set to merge this PR. I will also meanwhile pull this code and run it against our internal test suite to make sure that nothing breaks because of it. Will keep you posted on the results and if at all we would be required to make more changes.

Thanks!

File: src/yb/tools/yb-admin-test.cc [hunk 16]
```diff
@@ -538,6 +538,21 @@ TEST_F(AdminCliTest, TestGetClusterLoadBalancerState) {
...
 
+TEST_F(AdminCliTest, TestModifyPlacementPolicy) {
+  BuildAndStart();
+
+  // Modify the cluster placement policy to consist of 2 zones.
+  ASSERT_OK(CallAdmin("modify_placement_info", "c.r.z0,c.r.z1:2,c.r.z0:2", 5, ""));
+
+  auto output = ASSERT_RESULT(CallAdmin("get_universe_config"));
+
+  std::string expected_placement_blocks = "[{\"cloudInfo\":{\"placementCloud\":\"c\",\"placementRegion\":\"r\","
```
Comment (sanketkedia) [inline src/yb/tools/yb-admin-test.cc:549]: This line is > 100 characters. Our code style is that we generally keep lines within 100 characters. Hence break it down.

File: src/yb/tools/yb-admin_client.cc [hunk 17]
```diff
@@ -1729,7 +1729,7 @@ Status ClusterAdminClient::ModifyPlacementInfo(
   if (placement_info_split.size() < 1) {
     return STATUS(InvalidCommand, "Cluster config must be a list of "
     "placement infos seperated by commas. "
-    "Format: 'cloud1.region1.zone1,cloud2.region2.zone2,cloud3.region3.zone3 ..."
+    "Format: 'cloud1.region1.zone1:[min_replica_count1],cloud2.region2.zone2[min_replica_count2] ..."
```
Comment (sanketkedia) [inline src/yb/tools/yb-admin_client.cc:1732]: This line is > 100 characters.

File: src/yb/tools/yb-admin_client.cc [hunk 18]
```diff
@@ -1738,10 +1738,33 @@ Status ClusterAdminClient::ModifyPlacementInfo(
...
   std::unordered_map<std::string, int> placement_to_min_replicas;
-  for (int iter = 0; iter < placement_info_split.size(); iter++) {
-    placement_to_min_replicas[placement_info_split[iter]]++;
+  for (const auto& placement_block : placement_info_split) {
+    std::vector<std::string> placement_info_min_replica_split = strings::Split(
+        placement_block, ":", strings::AllowEmpty());
+
+    if(placement_info_min_replica_split.size() == 0 ||
+        placement_info_min_replica_split.size() > 2) {
+      return STATUS(InvalidCommand, "Each placement info must have at most 2 values separated by a colon. "
```
Comment (sanketkedia) [inline src/yb/tools/yb-admin_client.cc:1751]: This line is also > 100 characters

File: src/yb/tools/yb-admin_client.cc [hunk 19]
```diff
@@ -1738,10 +1738,33 @@ Status ClusterAdminClient::ModifyPlacementInfo(
...
+    if(placement_info_min_replica_split.size() == 2) {
+      placement_min_replica_count = VERIFY_RESULT(CheckedStoi(placement_info_min_replica_split[1]));
+    }
+
+    total_min_replica_count += placement_min_replica_count;
+    placement_to_min_replicas[placement_target] += placement_min_replica_count;
+  }
+
+  if (total_min_replica_count > replication_factor) {
+      return STATUS(InvalidCommand, "replication_factor should be greater than or equal to the total of replica counts specified in placement_info.");
```
Comment (sanketkedia) [inline src/yb/tools/yb-admin_client.cc:1767]: Same. Line is > 100

**Review by sanketkedia — CHANGES_REQUESTED** [review]
@NitronR once I pulled in your changes and ran lint against it, it showed several warnings. I have mentioned them as comments.

**Comment by sanketkedia** [comment]
@NitronR also pasting a link to our coding style guidelines for your reference. https://docs.yugabyte.com/latest/contribute/core-database/coding-style/

**Comment by NitronR** [comment]
@sanketkedia thanks for checking lint. Have pushed the updates. Can you point me to a resource that have steps for performing linting?

**Comment by NitronR** [comment]
@sanketkedia thanks for updating the PR and the detailed review comments 👍
